### PR TITLE
Add svd for real Symmetric and Hermitian matrices

### LINKED ISSUE #628: No svd method for Hermitian matrix
There is currently no method for `svd` with a hermitian matrix.
```julia
julia> using LinearAlgebra

julia> x = Hermitian(rand(5,5))
5×5 Hermitian{Float64,Array{Float64,2}}:
 0.0584939  0.45097   0.73698    0.945566  0.332148
 0.45097    0.91507   0.272532   0.514991  0.946693
 0.73698    0.272532  0.0604401  0.635022  0.906136
 0.945566   0.514991  0.635022   0.352346  0.5379
 0.332148   0.946693  0.906136   0.5379    0.484081

julia> svd(x)
ERROR: MethodError: no method matching svd(::Hermitian{Float64,Array{Float64,2}})
Closest candidates are:
  svd(::BitArray{2}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/bitarray.jl:90
  svd(::LinearAlgebra.AbstractTriangular) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/triangular.jl:2471
  svd(::Union{DenseArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2}, ReinterpretArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, ReshapedArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, AbstractCartesianIndex},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}, ::Union{DenseArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2}, ReinterpretArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, ReshapedArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, AbstractCartesianIndex},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}) where T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64} at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/svd.jl:281
  ...
Stacktrace:
 [1] top-level scope at none:0
```

```julia
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 12
  JULIA_EDITOR = nvim
```

### LINKED ISSUE #31986: No svd method for Hermitian matrix
There is currently no method for `svd` with a hermitian matrix.
```julia
julia> using LinearAlgebra

julia> x = Hermitian(rand(5,5))
5×5 Hermitian{Float64,Array{Float64,2}}:
 0.0584939  0.45097   0.73698    0.945566  0.332148
 0.45097    0.91507   0.272532   0.514991  0.946693
 0.73698    0.272532  0.0604401  0.635022  0.906136
 0.945566   0.514991  0.635022   0.352346  0.5379
 0.332148   0.946693  0.906136   0.5379    0.484081

julia> svd(x)
ERROR: MethodError: no method matching svd(::Hermitian{Float64,Array{Float64,2}})
Closest candidates are:
  svd(::BitArray{2}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/bitarray.jl:90
  svd(::LinearAlgebra.AbstractTriangular) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/triangular.jl:2471
  svd(::Union{DenseArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2}, ReinterpretArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, ReshapedArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, AbstractCartesianIndex},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}, ::Union{DenseArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2}, ReinterpretArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, ReshapedArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64},2,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, AbstractCartesianIndex},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}) where T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64} at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/svd.jl:281
  ...
Stacktrace:
 [1] top-level scope at none:0
```

```julia
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 12
  JULIA_EDITOR = nvim
```

### PR DESCRIPTION
**Pull request by dkarrasch**
Closes JuliaLang/LinearAlgebra.jl#628.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
If we could get a quick review from some committer with linalg expertise, this seems straightforward enough and it would be nice to get it merged.

**Comment by dkarrasch** [comment]
Oh, if it's really urgent, I think I can help here 😄 . I ended up using exactly @stevengj 's suggestion https://github.com/JuliaLang/julia/issues/31986#issuecomment-491925933 after the discussion in #31986. Should have mentioned that earlier.

And yes, since `svd(::Hermitian)`  errors on all 1.x versions, but this PR does not add a completely new functionality, I'd say this could be backported...?

**Comment by stevengj** [comment]
It's a new feature, not a bugfix, so it shouldn't be backported.

**Comment by StefanKarpinski** [comment]
> but this PR does not add a completely new functionality

I can see how it would seem that way since we already had SVD and Symmetric and Hermitian matrix types. However, adding previously unimplemented methods is adding new functionality. The general way to think about it is that patch releases should be both forward and backwards compatible with each other except for bugs: i.e. if your application works with Julia version x.y.z₁ then it should work with x.y.z₂ and vice versa unless there is a bug in one of those versions that your application specifically hits. Backporting this would violate that since someone could write code for Julia 1.1.2 that uses these `svd` methods which would not work on Julia 1.1.0 or 1.1.1 (which is about to be released).

**Comment by dkarrasch** [comment]
I see, thanks @StefanKarpinski for the explanation!

**Comment by StefanKarpinski** [comment]
@stevengj, could you review?

File: stdlib/LinearAlgebra/test/svd.jl
Comment (stevengj) [inline stdlib/LinearAlgebra/test/svd.jl]: Maybe combine these two cases:
```jl
@testset "singular value decomposition of Hermitian/real-Symmetric" begin
    for T in eltya <: Real ? (Symmetric, Hermitian) : (Hermitian,)
        usv = svd(T(asym))
        @test ...
    end
end

File: stdlib/LinearAlgebra/test/svd.jl
Comment (stevengj) [inline stdlib/LinearAlgebra/test/svd.jl]: Also seems like this should be outside the `for (a, a2)` loop since it only uses `asym`

Comment (dkarrasch) [reply stdlib/LinearAlgebra/test/svd.jl]: Good catch! Actually, that loop is mysterious. `a2` is never used in it. I should still move the test outside, but should I also simplify that loop? Or leave it until someone guesses what `a2` was meant to be good for and includes it in tests?

Comment (stevengj) [reply stdlib/LinearAlgebra/test/svd.jl]: Simplifying that loop would be good.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/test/svd.jl]: There are a couple of other unused variables. I wonder about the following: after setting `Random.seed!(1234321)`, and then calling
```julia
areal = randn(n,n)/2
aimg  = randn(n,n)/2
a2real = randn(n,n)/2 # unused
a2img  = randn(n,n)/2 # unused
```
the latter two are not used, anywhere. If I remove them, does that change the random numbers created in subsequent `rand*` calls? Probably yes. But is that an issue? Should I leave them for test consistency? I don't think that we rely here on that the random numbers in future tests are exactly the same as they used to be, but I'm not sure how strictly that is handled.

Comment (StefanKarpinski) [reply stdlib/LinearAlgebra/test/svd.jl]: > Should I leave them for test consistency?

No, it's ok to change them, but if tests start failing it may require using a different seed or loosening some approximate equality tests. The main reason for using a fixed seed is so that we don't get random failures during CI, but if the bounds are good then most seeds should be fine.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/test/svd.jl]: Tests pass locally.

**Review by stevengj — APPROVED** [review]

**Pull request merged**
